### PR TITLE
Improve performance of IamDataFrame initialization

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- ~Tests Added~ -> no functional changes, existing tests should cover all cases where something could go wrong
- [x] Name of contributors Added to AUTHORS.rst
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR implement faster and less memory-intensive initialization of an IamDataframe by doing the sorting and validation on the indexed pd.Series instead of the DataFrame. Based on memory-profiling work by @phackstock.

### CODE REVIEW

**Comment by danielhuppmann** [comment]
Thanks to @sandrinecharousset who reported an "out-of-memory" issue when loading large files as IamDataFrame - which motivated @phackstock and me to dig a bit into the performance...

File: pyam/utils.py [hunk 1]
```diff
@@ -329,21 +329,28 @@ def convert_r_columns(c):
...
     idx_cols = index + REQUIRED_COLS + [time_col] + extra_cols
-    rows = df[idx_cols].duplicated()
+    df = df.set_index(idx_cols).value
+
+    rows = df.index.duplicated()
     if any(rows):
-        _raise_data_error("duplicate rows in `data`", df.loc[rows, idx_cols])
+        _raise_data_error(
+            "Duplicate rows in `data`", df[rows].index.to_frame(index=False)
+        )
```
Comment (phackstock) [inline pyam/utils.py:349]: In analogy to `del null_row` we could also delete rows, saving a little bit more memory
```suggestion
    del rows
```

**Review by phackstock — COMMENTED** [review]
Minus the one question below, very nice PR with impactful changes.
Running `pyam.IamDataFrame(csv_data)` with half a GB of csv data as a benchmark gives the following results:

| | peak memory usage (GB) | execution time (s) |
| - | - | - |
| pre PR #579 | 8 | 105 |
| with PR #579 | 7.1 (13% less) | 76 (28% faster)|

File: pyam/utils.py [hunk 2]
```diff
@@ -329,21 +329,28 @@ def convert_r_columns(c):
     df.loc[df.unit.isnull(), "unit"] = ""
 
     # verify that there are no nan's left (in columns)
-    null_rows = df.isnull().values
+    null_rows = df.isnull().T.any()
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/utils.py:332]: This is not duplicated, because `df.isnull()` returns a matrix of bools - the first `T.any()` reduces this to "is there any null in a row", the next line asks "is there any null in any row" - and prints out the offending row(s) if so.

[the remark this answers is not in the document]
Comment (phackstock) [reply pyam/utils.py:332]: Ah ok I see, then looks good to me.

**Review by phackstock — APPROVED** [review]
Thanks for adding me to the authors list. 
PR looks good to me.

**Comment by danielhuppmann** [comment]
@gidden @znicholls @Rlamboll @coroa, any thoughts? If there are no objections, I'll merge next week.

**Comment by Rlamboll** [comment]
Looks good to me and it's always good to speed things up a little!

File: pyam/utils.py [hunk 3]
```diff
@@ -329,21 +329,28 @@ def convert_r_columns(c):
...
     if null_rows.any():
-        _raise_data_error("empty cells in `data`", df.loc[null_rows])
+        _raise_data_error("Empty cells in `data`", df.loc[null_rows])
+    del null_rows
```
Comment (znicholls) [inline pyam/utils.py:335]: Why is this needed?

File: pyam/utils.py [hunk 4]
```diff
@@ -329,21 +329,28 @@ def convert_r_columns(c):
...
-    if any(rows):
-        _raise_data_error("duplicate rows in `data`", df.loc[rows, idx_cols])
+    df = df.set_index(idx_cols).value
 
+    rows = df.index.duplicated()
+    if any(rows):
+        _raise_data_error(
+            "Duplicate rows in `data`", df[rows].index.to_frame(index=False)
+        )
+    del rows
```
Comment (znicholls) [inline pyam/utils.py:349]: Why is this needed

**Review by znicholls — COMMENTED** [review]
I don't understand why the del statements are there (is someone running into memory issues?), otherwise looks good

**Comment by znicholls** [comment]
Right I've now read the whole thing so have answered my own question, all good

**Comment by danielhuppmann** [comment]
Thanks @phackstock for the effort, thanks @znicholls & @Rlamboll for taking a look at the implementation!
